### PR TITLE
Update predict functions to align with pytorch lightning API

### PR DESCRIPTION
**Pull request by bw4sz**
In response to #407, the goal of this PR is to remove any .to(CUDA) statements and let pytorch lightning manage devices. I sense that this will require updating the dataloaders for predict_image and predict_tile. I am testing locally on CPU and on HPC GPU.

### CODE REVIEW

**Comment by bw4sz** [comment]
I will also attack #404 in trying to compare the timing of this PR against predict_tile in main.

**Comment by bw4sz** [comment]
This passes, the same error that is unrelated in the other PR. To do items 

- [] Documentation of pytorch lightning options and integration "How do I make this faster" page
- [] Timings on cpu and gpu for a large predict tile to assess speed improvements.

**Comment by bw4sz** [comment]
Here are timings for a 128 mb tile, about 6000 trees, general code is 

```
from deepforest import main
import cProfile, pstats

def run():
    m = main.deepforest()
    m.config["workers"] = 0    
    print(m.config)
    m.use_release()
    boxes = m.predict_tile("2018_TEAK_3_315000_4094000_image_crop.tif")
    
    return boxes


prof = cProfile.Profile()
stats = prof.run("run()")
stats = pstats.Stats(stats)
stats.sort_stats(pstats.SortKey.TIME).print_stats(10)
```

```
(base) [b.weinstein@login3 DeepForest]$ srundev -p gpu --gpus=1 --mem 40GB --cpus-per-task 10
```
## This PR

batch size ==1
0 workers: 11.96s
1 workers:  13.86s
10 workers: 14s

batch size = 32
0 workers: 9.44s
1 workers: 9.9s
2 workers: 99s
10 workers: 11.61

## main branch

Main branch has no batch prediction, all tests are batch size =1 
0 workers: 15.572s
1 workers: 13.14s
2 workers: 13.44s
10 workers: 12.86s

The batch prediction and predict_dataloader both add time improvements. Overall its about 33% faster, and I expect those gains to be clearer at larger scales. One challenge I see, and have seen elsewhere is that workers = 0, which is no multi-processing and runs on main thread, is faster than workers > 0. 

https://discuss.pytorch.org/t/guidelines-for-assigning-num-workers-to-dataloader/813/9

I haven't played with pin_memory, its very hard to anticipate all the GPU configs that users will encounter. I think we keep workers ==1 as config default,  a slight performance hit, but safer from memory allocation.

**Comment by ethanwhite** [comment]
I started a review, but won't be able to finish it until the current grant proposal is submitted. I'm going to post the partial review so that at least you have access to it. If you still want me to look at this after the proposal ships please remind me then.

**Review by ethanwhite — CHANGES_REQUESTED** [review]

File: deepforest/dataset.py [hunk 1]
```diff
@@ -133,3 +135,45 @@ def __getitem__(self, idx):
             # Mimic the train augmentation
             converted = self.image_converter(image=image)
             return converted["image"]
+
+class TileDataset(Dataset):
+    def __init__(self, 
+                 tile: typing.Optional[np.ndarray]=None,
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/dataset.py:141]: yes, you read it right, I had copied and pasted from another function. Good catch. Corrected to 

```
class TileDataset(Dataset):
    def __init__(self, 
                 tile: typing.Optional[np.ndarray], 
                 preload_images: bool=False,
                 patch_size: int=400,
                 patch_overlap: float=0.05
                 ):
```

File: deepforest/main.py [hunk 2]
```diff
@@ -132,6 +128,8 @@ def create_trainer(self, logger=None, callbacks=[], **kwargs):
         if not self.config["validation"]["csv_file"] is None:
             if logger is not None:
                 lr_monitor = LearningRateMonitor(logging_interval='epoch')
+                iou_callback = iou_callback(self.config, every_n_epochs=self.config["validation"]["val_accuracy_interval"])
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/main.py:131]: Changed to 

```
                eval_callback = iou_callback(self.config, every_n_epochs=self.config["validation"]["val_accuracy_interval"])
```

File: deepforest/callbacks.py [hunk 3]
```diff
@@ -66,6 +62,49 @@ def log_images(self, pl_module):
...
+        self.every_n_epochs = every_n_epochs
+
+    def on_validation_epoch_end(self, trainer, pl_module):
+        if trainer.current_epoch % self.every_n_epochs == 0:
+            results = pl_module.evaluate(csv_file=self.config["validation"]["csv_file"],
+                                    root_dir=self.config["validation"]["root_dir"])
+            self.log("box_recall", results["box_recall"])
+            self.log("box_precision", results["box_precision"])
+
+            if not type(results["class_recall"]) == type(None):
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/callbacks.py:101]: Pandas dataframes are more annoying than that. 

```
import pandas as pd
a = pd.DataFrame()
if a: print("hi")

Traceback (most recent call last):
  Python Shell, prompt 3, line 1
    # Used internally for debug sandbox under external interpreter
  File "/Users/benweinstein/.conda/envs/DeepForest/lib/python3.9/site-packages/pandas/core/generic.py", line 1534, in __nonzero__
    raise ValueError(
builtins.ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

but this works

```
if not type(a) == type(None): print("hi")
hi
```

both work for the None condition

```
if a: print("hi")
if not type(a) == type(None): print("hi")
``

[the remark this answers is not in the document]
Comment (ethanwhite) [reply deepforest/callbacks.py:101]: Ah, yeah. So, we're just checking to see if it's a dataframe as opposed to `None`? I'd probably use:

```python
if isinstance(a, pd.DataFrame):
```

Or, if it's in our control, I'd force type consistency in the output by stopping `evaluate` from returning `None` for `results["class_recall"]`, have it pass an empty dataframe instead, and use:

```python
if not a.empty:
```

[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/callbacks.py:101]: I was looking at this, here is a situation where we get empty class recall

```
def test_evaluate_empty():
    m = main.deepforest()
    m.config["score_thresh"] = 0.8
    csv_file = get_data("OSBS_029.csv")
    root_dir = os.path.dirname(csv_file)
    results = m.evaluate(csv_file, root_dir, iou_threshold = 0.4)
    
    #Does this make reasonable predictions, we know the model works.
    assert np.isnan(results["box_precision"])
    assert results["box_recall"] == 0
    
    df = pd.read_csv(csv_file)
    assert results["results"].shape[0] == df.shape[0]
```
Here there is prediction

```
results["predictions"]
Empty DataFrame
Columns: [xmin, ymin, xmax, ymax, label, score, image_path]
Index: []
```

I think it makes sense to stop at that stage, if there are no predictions that pass the score threshold, just return None and a warning. Adding to evaluate.py

```
    if predictions.empty:
        warnings.warn("No predictions were empty, returning None")
        return None
```

[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/callbacks.py:101]: Nevermind, I changed my mind, the evaluate accuracy of no predictions is 0, its not one. I think it makes more sense to return the dataframe as is, but with each line of recall as 0.

[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/callbacks.py:101]: In the end I left it as it was and just adopted the suggested.

```
if isinstance(a, pd.DataFrame):
```
in callbacks.py. I was worried about making a change that I could imagine be breaking just to adopt a new syntax elsewhere. I can return to the question of whether null predictions should have per class recall of None or 0 elsewhere.

[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/callbacks.py:101]: Made a new issue https://github.com/weecology/DeepForest/issues/414

**Comment by bw4sz** [comment]
Let's get this merged and then we can deal separately with the azure errors.

**Comment by ethanwhite** [comment]
> Let's get this merged and then we can deal separately with the azure errors.

Agreed. I'll finish looking this over as best I can in the next few minutes and then get one more review with the existing bits I've accumulated. Once you look at those and let me know we're ready I'll push the button.

**Review by ethanwhite — CHANGES_REQUESTED** [review]
Looks like I'd stopped just a file away from the test changes a week ago. Sigh. Sorry for the delay. There are a few more things I'd already written up so going ahead and posting this. Feel free to ignore things that aren't important (like the conditional switch).

File: deepforest/predict.py [hunk 4]
```diff
@@ -73,212 +66,58 @@ def predict_image(model,
...
-    If "label" column is present, these are assumed to be annotations and will be plotted in a different color than predictions
+def mosiac(boxes, windows, use_soft_nms=False, sigma=0.5, thresh=0.001, iou_threshold=0.1):
+    # transform the coordinates to original system
+    
+    for index, window_boxes in enumerate(boxes):
+        xmin, ymin, xmax, ymax = windows[index].getRect()
+        window_boxes.xmin = window_boxes.xmin + xmin
+        window_boxes.xmax = window_boxes.xmax + xmin
+        window_boxes.ymin = window_boxes.ymin + ymin
+        window_boxes.ymax = window_boxes.ymax + ymin
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/predict.py:77]: removed.

File: deepforest/predict.py [hunk 5]
```diff
@@ -73,212 +66,58 @@ def predict_image(model,
...
+        # their intersection-over-union (IoU).
+        bbox_left_idx = nms(boxes=boxes,
+                            scores=scores,
+                            iou_threshold=iou_threshold)
+    else:
+        # Performs soft non-maximum suppression (soft-NMS) on the boxes.
+        bbox_left_idx = soft_nms(boxes=boxes,
+                                 scores=scores,
+                                 sigma=sigma,
+                                 thresh=thresh)
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/predict.py:100]: done.

File: deepforest/predict.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bw4sz) [reply deepforest/predict.py:77]: the boxes list is being modified in place. Perhaps not the cleanest way, but the windows object is alittle awkward. 

```
boxes[3].xmin
0     231.347656
1     161.749283
2     114.964119
3     154.518982
4     195.182327
5      77.563629
6     279.870026
7     218.403671
8      89.309868
9      95.011116
10    189.664459
11    265.632446
12    186.969330
13     96.969086
14      1.647566
15    133.053711
16    175.091278
17    273.805359
18     66.780685
19    226.849945
20    128.855423
21    137.390427
22    288.996460
23    292.613800
24    284.085510
25     18.438725
26     61.180923
27    136.890686
28     51.033783
29    228.117874
30      0.000000
31     62.546310
32     60.722111
33     54.490486
34     30.509073
35      0.000000
36    141.686020
37      0.000000
38     29.767782
39    247.072678
40    175.313080
41     72.166641
42    276.608887
43    168.369614
44    293.687439
45    205.511444
46      0.140940
47    287.159180
48    164.969666
Name: xmin, dtype: float32
for index, window_boxes in enumerate(boxes):
    xmin, ymin, xmax, ymax = windows[index].getRect()
    window_boxes.xmin = window_boxes.xmin + xmin
    window_boxes.xmax = window_boxes.xmax + xmin
    window_boxes.ymin = window_boxes.ymin + ymin
    window_boxes.ymax = window_boxes.ymax + ymin

boxes[3].xmin
0     331.347656
1     261.749268
2     214.964111
3     254.518982
4     295.182312
5     177.563629
6     379.870026
7     318.403687
8     189.309875
9     195.011108
10    289.664459
11    365.632446
12    286.969330
13    196.969086
14    101.647568
15    233.053711
16    275.091278
17    373.805359
18    166.780685
19    326.849945
20    228.855423
21    237.390427
22    388.996460
23    392.613800
24    384.085510
25    118.438721
26    161.180923
27    236.890686
28    151.033783
29    328.117859
30    100.000000
31    162.546310
32    160.722107
33    154.490479
34    130.509079
35    100.000000
36    241.686020
37    100.000000
38    129.767776
39    347.072693
40    275.313080
41    172.166641
42    376.608887
43    268.369629
44    393.687439
45    305.511444
46    100.140938
47    387.159180
48    264.969666
Name: xmin, dtype: float32
```

[the remark this answers is not in the document]
Comment (ethanwhite) [reply deepforest/predict.py:77]: Oh, I see. Relying on the aliasing is definitely confusing (at least for me). What about being explicit:

```python
    for index, _ in enumerate(boxes): # or for index in range(len(boxes)), which will be a bit faster
        xmin, ymin, xmax, ymax = windows[index].getRect()
        boxes[index].xmin += xmin
        boxes[index].xmax +=  xmin
        boxes[index].ymin +=  ymin
        boxes[index].ymax +=  ymin
```

**Comment by ethanwhite** [comment]
The new test failures are from the Pandas 2.0 release this week. I'll open a new issue.

I'll merge once the change to the loop with the aliasing is in or if you tell  me you want to keep it the way it is.

**Review by ethanwhite — APPROVED** [review]

**Comment by ethanwhite** [comment]
Current errors seem related to #422 so merging so that we can get all of the bits and pieces in.

**Pull request merged**